### ISSUE TITLE
Can't compile anything to Armv7 host. Not identifying host CPU correctly?

### ISSUE THREAD

**Issue by arm32-zig** [ziglang/zig#5110]
I downloaded the 0.6 binary for zig-linux-armv7a-0.6.0, extracted it and tried to compile a simple program created by zig init-exe. I ran into the following error.

```
<inline asm>:1:2: error: instruction requires: armv6k

        yield
        ^

LLVM ERROR: Error parsing inline asm<inline asm>:1:2: error: instruction requires: armv6k

        yield
        ^
LLVM ERROR: Error parsing inline asm
```

I'm running this on an odoid-xu3 which has a armv71 cortex_a7 and cortex_a15 under linux.
That cpu should support a yield instruction.

I tried using -mcpu to target the cortex_a7 explicitly, but I got the same error.
Figuring there might be an error in the pre-built binary I downloaded the boot-strap source from
zig-bootstrap-0.6.0.tar.xz and tried to build the compiler with the following commands
```
build -j1 arm-linux-musleabihf native
```
The build failed right after it finished building zig0. Here is the error.
```
Scanning dependencies of target zig_build_libstage2

<inline asm>:1:2: error: instruction requires: armv6k
        yield
        ^
 LLVM ERROR: Error parsing inline asm
```
Same thing as before. It appears that the self-hosted compiler can't generate any code b/c it
doesn't understand the cpu features of the host correctly. There is no good reason why it should
not be able to run a yield instruction. ARM developer documentation says it is supported.

**Comment by LemonBoy** [comment]
Can you paste the output of `cat /proc/cpuinfo` and `zig --verbose-llvm-cpu-features build-exe <your usual flags>` ?

**Comment by arm32-zig** [comment]
Sure thing.
cat /proc/cpuinfo
```
processor	: 0
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 18.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xc07
CPU revision	: 3

processor	: 1
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 18.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xc07
CPU revision	: 3

processor	: 2
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 18.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xc07
CPU revision	: 3

processor	: 3
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 18.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xc07
CPU revision	: 3

processor	: 4
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 90.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0xc0f
CPU revision	: 3

processor	: 5
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 90.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0xc0f
CPU revision	: 3

processor	: 6
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 90.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0xc0f
CPU revision	: 3

processor	: 7
model name	: ARMv7 Processor rev 3 (v7l)
BogoMIPS	: 90.00
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0xc0f
CPU revision	: 3

Hardware	: ODROID-XU3
Revision	: 0100
Serial		: 0000000000000000
```
zig --verbose-llvm-cpu-features build-exe src/main.zig
```
./zig --verbose-llvm-cpu-features build-exe src/main.zig
name=main triple=arm-unknown-linux-gnueabihf
name=main target_specific_cpu_args=generic
name=main target_specific_features=-32bit,-8msecext,-a76,-aclass,-acquire-release,-aes,-avoid-movs-shop,-avoid-partial-cpsr,-cheap-predicable-cpsr,-crc,-crypto,+d32,-db,-dfb,-disable-postra-scheduler,-dont-widen-vmovs,-dotprod,-dsp,-execute-only,-expand-fp-mlx,-exynos,+fp16,-fp16fml,+fp64,-fp-armv8,-fp-armv8d16,-fp-armv8d16sp,-fp-armv8sp,-fpao,+fpregs,-fpregs16,+fpregs64,-fullfp16,-fuse-aes,-fuse-literals,-v4t,-v5t,-v5te,-v6,-v6k,-v6m,-v6t2,-v7,-v7clrex,-v8.1a,-v8.1m.main,-v8.2a,-v8.3a,-v8.4a,-v8.5a,-v8,-v8m,-v8m.main,+hwdiv,+hwdiv-arm,-iwmmxt,-iwmmxt2,-lob,-long-calls,-loop-align,-m3,-mclass,-mp,-muxed-units,-mve,-mve.fp,-mve1beat,-mve2beat,-mve4beat,-nacl-trap,+neon,-neon-fpmovs,-neonfp,-no-branch-predictor,-no-movt,-no-neg-immediates,-noarm,-nonpipelined-vfp,-perfmon,-prefer-ishst,-prefer-vmovsr,-prof-unpr,-r4,-ras,-rclass,-read-tp-hard,-reserve-r9,-ret-addr-stack,-sb,-sha2,-slow-fp-brcc,-slow-load-D-subreg,-slow-odd-reg,-slow-vdup32,-slow-vgetlni32,-slowfpvfmx,-slowfpvmlx,-soft-float,-splat-vfp-neon,-strict-align,-swift,-thumb2,-thumb-mode,-trustzone,-use-misched,-armv2,-armv2a,-armv3,-armv3m,-armv4,-armv4t,-armv5t,-armv5te,-armv5tej,-armv6,-armv6j,-armv6k,-armv6kz,-armv6-m,-armv6s-m,-armv6t2,-armv7-a,-armv7e-m,-armv7k,-armv7-m,-armv7-r,-armv7s,-armv7ve,-armv8-a,-armv8-m.base,-armv8-m.main,-armv8-r,-armv8.1-a,-armv8.1-m.main,-armv8.2-a,-armv8.3-a,-armv8.4-a,-armv8.5-a,+vfp2,+vfp2sp,+vfp3,+vfp3d16,+vfp3d16sp,+vfp3sp,+vfp4,+vfp4d16,+vfp4d16sp,+vfp4sp,-virtualization,-vldn-align,-vmlx-forwarding,-vmlx-hazards,-wide-stride-vfp,-xscale,-zcz
LLVM Emit Output...<inline asm>:1:2: error: instruction requires: armv6k
        yield
        ^
LLVM ERROR: Error parsing inline asm

```

**Comment by LemonBoy** [comment]
Zig isn't able to detect the native CPU type on anything but X86 (it tries to use `generic+v7a` in that case but `crossTargetToTarget` undoes everything and returns a simple `generic`) and the complex machinery that's meant to pick the correct target flags fails to do so. What you get is a non-existing target (`-v4t,-v5t,-v5te,-v6,-v6k,-v6m,-v6t2,-v7,-v7clrex,-v8.1a,-v8.1m.main,-v8.2a,-v8.3a,-v8.4a,-v8.5a,-v8,-v8m,-v8m.main`) and lots of problems.

Using the `mcpu` switch works fine here.

**Comment by arm32-zig** [comment]
I tried `zig -mcpu generic+v7a build-exe src/main.zig` and it worked!

Trying `zig  build -mcpu generic+v7a` doesn't work though. I dont' know if the build.zig file overrides
the command line option or not. Either way, it would be good if some kind of warning or error were
issued so future users don't think the mcpu option is actually being heeded.

I'll stick with the workaround for now. Thanks for your help!

**Comment by andrewrk** [comment]
Zig uses LLVM's CPU feature detection code on non x86 hosts, however it incorrectly does not detect the subarch and convert that to a CPU feature flag.

I think the path forward here is to improve the self hosted CPU detection code to support ARM. #4591

**Comment by andrewrk** [comment]
Fixed a long time ago in #8671.
